packages/apps/src/app.events.ts

```diff
@@ -8,23 +8,51 @@ import {
   IErrorEvent,
   IEvents,
 } from './events';
-import { ISender } from './types';
+import { IPlugin, IPluginWithEvents, ISender } from './types';
+
+type UnionToIntersection<U> =
+  (U extends any ? (k: U) => void : never) extends (k: infer I) => void ? I : never;
+
+type PluginEvents<T> =
+  T extends IPluginWithEvents<infer Events> ? Events : {};
+
+type MergePluginEventMaps<TPlugins> =
+  UnionToIntersection<
+    TPlugins extends readonly unknown[] ? PluginEvents<TPlugins[number]> : PluginEvents<TPlugins>
+  >;
+
+// Prioritizes the keys of IEvents, and merges all the Events for the Plugins
+export type AppEvents<TPlugins> = {
+  [K in keyof IEvents | keyof MergePluginEventMaps<TPlugins>]:
+  K extends keyof IEvents
+  ? IEvents[K]
+  : K extends keyof MergePluginEventMaps<TPlugins>
+  ? MergePluginEventMaps<TPlugins>[K]
+  : never;
+};
```

Comment:
move all this to a diff file?

Reply:
These are great, I would suggest we add some more verbose comments to the `AppEvents` since it gets exported.